crates/holochain/src/core/workflow/app_validation_workflow.rs

```diff
@@ -658,70 +643,75 @@ where
         ValidateResult::Valid => Ok(Outcome::Accepted),
         ValidateResult::Invalid(reason) => Ok(Outcome::Rejected(reason)),
         ValidateResult::UnresolvedDependencies(UnresolvedDependencies::Hashes(hashes)) => {
-            // This is the base case where we've been recursing and start seeing
-            // all the same hashes unresolved that we already tried to fetch.
-            // At this point we should just give up on the inline recursing and
-            // let some future background task attempt to fetch these hashes
-            // again. Hopefully by then the hashes are fetchable.
-            // 20 is a completely arbitrary max recursion depth.
-            if recursion_depth > 20 || hashes.iter().all(|hash| fetched_deps.contains(hash)) {
-                Ok(Outcome::AwaitingDeps(hashes))
-            } else {
-                let in_flight = hashes.into_iter().map(|hash| async {
-                    let cascade_workspace = workspace_read.clone();
-                    let cascade = CascadeImpl::from_workspace_and_network(
-                        &cascade_workspace,
-                        network.clone(),
-                    );
-                    cascade
-                        .fetch_record(hash.clone(), NetworkGetOptions::must_get_options())
-                        .await?;
-                    Ok(hash)
-                });
-                let results: Vec<_> = futures::stream::iter(in_flight)
-                    // 10 is completely arbitrary.
-                    .buffered(10)
-                    .collect()
-                    .await;
-                let results: AppValidationResult<Vec<_>> = results.into_iter().collect();
-                for hash in results? {
-                    fetched_deps.insert(hash);
+            tracing::debug!("unresolved dependencies: hashes {hashes:?}");
+            // fetch all missing hashes in the background without awaiting them
+            let cascade_workspace = workspace_read.clone();
+            let cascade =
+                CascadeImpl::from_workspace_and_network(&cascade_workspace, network.clone());
+            // build a collection of futures to fetch the individual missing
+            // hashes
+            let fetches = hashes.clone().into_iter().map(move |hash| {
+                let cascade = cascade.clone();
+                // keep track of which dependencies are being fetched to
+                // prevent multiple fetches of the same hash
+                let fetched_dependencies = fetched_dependencies.clone();
+                async move {
```

Comment:
You could make this a filter_map and do the hash set insertion outside of the future, and only create a new fetch future if it's new_dependency. This could create a bunch of empty futures needlessly.

Reply:
Great catches, fixed